### PR TITLE
Import wind downscaling

### PR DESCRIPTION
**Pull request by gavinevans**
Edits in response to review comments.

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
FYI, I've rebased the import_wind_downscaling branch and then done some conflict resolution online magic on this PR

File: lib/improver/wind_downscaling.py [hunk 1]
```diff
@@ -573,7 +575,7 @@ def _delta_height(self):
         delt_z[self.hcmask] = self.pporo[self.hcmask]-self.modoro[self.hcmask]
         return delt_z
 
-    def _do_rc_hc_all(self, hgrid, uorig):
+    def do_rc_hc_all(self, hgrid, uorig):
```
Comment (benfitzpatrick) [inline lib/improver/wind_downscaling.py:577]: Did we want this public?

File: lib/improver/wind_downscaling.py [hunk 2]
```diff
@@ -646,19 +652,6 @@ def __init__(self, a_over_s_cube, sigma_cube, pporo_cube,
...
-        if hasattr(sigma_cube, "operation"):
-            sigma_cube = sigma_cube.operation()
-        if hasattr(pporo_cube, "operation"):
-            pporo_cube = pporo_cube.operation()
-        if hasattr(modoro_cube, "operation"):
-            modoro_cube = modoro_cube.operation()
-        if hasattr(z0_cube, "operation"):
-            z0_cube = z0_cube.operation()
-        if hasattr(height_levels_cube, "operation"):
-            height_levels_cube = height_levels_cube.operation()
```
Comment (benfitzpatrick) [inline lib/improver/wind_downscaling.py:660]: Thanks for this

File: lib/improver/wind_downscaling.py [hunk 3]
```diff
@@ -711,33 +697,25 @@ def find_coord_names(self, cube):
         clist = set([cube.coords()[iii].name() for iii in
                      range(len(cube.coords()))])
         try:
-            xname = list(clist.intersection(self.xcoordnames))[0]
-        except IndexError:
-            xname = None
-        except Exception as exc:
+            xname = cube.coord(axis="x").name()
```
Comment (benfitzpatrick) [inline lib/improver/wind_downscaling.py:699]: Nice

File: lib/improver/wind_downscaling.py [hunk 4]
```diff
@@ -933,58 +915,56 @@ def check_wind_ancil(self, xwp, ywp):
...
-                                      ordered=False):
+        hld = self.find_heightgrid(input_cube)
+        for time_slice in input_cube.slices_over("time"):
             if np.isnan(time_slice.data).any() or (time_slice.data < 0.).any():
                 msg = ('{} has invalid wind data')
                 raise ValueError(msg.format(time_slice.coord(self.t_name)))
             rc_hc = copy.deepcopy(time_slice)
-            rc_hc.data = roughness_correction._do_rc_hc_all(hld,
-                                                            time_slice.data)
+            rc_hc.data = roughness_correction.do_rc_hc_all(
```
Comment (benfitzpatrick) [inline lib/improver/wind_downscaling.py:955]: Ah I see about the public

File: lib/improver/wind_downscaling.py [hunk 5]
```diff
@@ -933,58 +915,56 @@ def check_wind_ancil(self, xwp, ywp):
...
             z0_data = self.z_0.data
         roughness_correction = RoughnessCorrectionUtilities(
             self.a_over_s.data, self.sigma.data, z0_data, self.pp_oro.data,
             self.model_oro.data, self.ppres, self.modres)
         self.check_wind_ancil(xwp, ywp)
-        hld = self.find_heightgrid(wind)
-        for time_slice in wind.slices([self.y_name, self.x_name, self.z_name],
-                                      ordered=False):
+        hld = self.find_heightgrid(input_cube)
+        for time_slice in input_cube.slices_over("time"):
```
Comment (benfitzpatrick) [inline lib/improver/wind_downscaling.py:950]: Lovely

File: lib/improver/wind_downscaling.py [hunk 6]
```diff
@@ -860,20 +843,14 @@ def find_coord_order(self, mcube):
...
-            except Exception:
-                pass
-        return xpos, ypos, zpos, tpos
+        coord_names = [self.x_name, self.y_name, self.z_name, self.t_name]
+        positions = [np.nan, np.nan, np.nan, np.nan]
+        for coord_index, coord_name in enumerate(coord_names):
+            if mcube.coords(coord_name, dim_coords=True):
+                coord_dimension = mcube.coord_dims(coord_name)
+                coord_dimension = coord_dimension[0]
+                positions[coord_index] = coord_dimension
```
Comment (benfitzpatrick) [inline lib/improver/wind_downscaling.py]: Collapse to either 1 or 2 lines here?

File: lib/improver/wind_downscaling.py (hunk 1 above)
Comment (benfitzpatrick) [reply lib/improver/wind_downscaling.py:577]: Retracted, saw where it was used below

**Comment by benfitzpatrick** [comment]
One comment about a 3 line block that I think could be 2 lines or 1, otherwise good.

File: lib/improver/wind_downscaling.py (hunk 6 above)
Comment (gavinevans) [reply lib/improver/wind_downscaling.py]: Collapsed to 1 line.

**Pull request merged**